File: core/src/main/scala/com/graphbrain/db/UserNode.scala
package com.graphbrain.db

import org.mindrot.jbcrypt.BCrypt
import java.math.BigInteger
import com.graphbrain.utils.RandUtils


case class UserNode(override val id: String,
                    username: String,
                    name: String,
                    email: String,
                    pwdhash: String,
                    role: String="",
                    session: String="",
                    sessionTs: Long= -1,
                    lastSeen: Long= -1,
                    override val summary: String="",
                    override val degree: Int = 0,
                    override val ts: Long = -1)
  extends Textual(id, summary, degree, ts) {

  def this(id: String, map: Map[String, String]) =
    this(id,
      map("username"),
      map("name"),
      map("email"),
      map("pwdhash"),
      map("role"),
      map("session"),
      map("sessionTs").toLong,
      map("lastSeen").toLong,
      map("summary"),
      map("degree").toInt,
      map("ts").toLong)

  override def extraMap = Map("username" -> username,
                              "name" -> name,
                              "email" -> email,
                              "pwdhash" -> pwdhash,
                              "role" -> role,
                              "session" -> session,
                              "sessionTs" -> sessionTs.toString,
                              "lastSeen" -> lastSeen.toString,
                              "summary" -> summary)

  override def setId(newId: String): Vertex = copy(id=newId)

  override def setDegree(newDegree: Int): Vertex = copy(degree=newDegree)

  override def setTs(newTs: Long): Vertex = copy(ts=newTs)

  def newSession = copy(session= new BigInteger(130, RandUtils.secRand).toString(32))

  def checkPassword(candidate: String) = BCrypt.checkpw(candidate, pwdhash)

  def checkSession(candidate: String) = session == candidate

  override def toString: String = name

  //override def updateSummary: Textual = UserNode(id, username, name, email, pwdhash, role, session, sessionTs, lastSeen, generateSummary)

  override def raw: String = {
    "type: " + "user<br />" +
    "username: " + username + "<br />" +
    "name: " + name + "<br />" +
    "role: " + role + "<br />" +
    "lastSeen: " + lastSeen + "<br />"
  }
}

object UserNode {
  def create(username: String, name: String, email: String, password: String, role: String) = {
    val id = ID.userIdFromUsername(username)
    val pwdhash = BCrypt.hashpw(password, BCrypt.gensalt())
    UserNode(id, username, name, email, pwdhash, role)
  }
}

File: core/src/main/java/com/graphbrain/db/UserNode.java
package com.graphbrain.db;

import org.mindrot.jbcrypt.BCrypt;
import java.math.BigInteger;
import java.util.Map;

import com.graphbrain.utils.RandUtils;


public class UserNode extends Vertex {

    private String username;
    private String name;
    private String email;
    private String pwdhash;
    private String role;
    private String session;
    private long sessionTs;
    private long lastSeen;

    @Override
    public VertexType type() {return VertexType.User;}

    public UserNode(String id,
        String username,
        String name,
        String email,
        String pwdhash,
        String role,
        String session,
        long sessionTs,
        long lastSeen,
        int degree,
        long ts) {

        super(id, degree, ts);
        this.username = username;
        this.name = name;
        this.email = email;
        this.pwdhash = pwdhash;
        this.role = role;
        this.session = session;
        this.sessionTs = sessionTs;
        this.lastSeen = lastSeen;
    }

    public UserNode(String id,
                    String username,
                    String name,
                    String email,
                    String pwdhash,
                    String role) {

        this(id, username, name, email, pwdhash, role, "", -1, -1, 0, -1);
    }

    @Override
    public Vertex copy() {
        return new UserNode(id, username, name, email, pwdhash, role, session, sessionTs, lastSeen, degree, ts);
    }

    public static UserNode create(String username, String name, String email, String password, String role) {
        String id = ID.userIdFromUsername(username);
        String pwdhash = BCrypt.hashpw(password, BCrypt.gensalt());
        return new UserNode(id, username, name, email, pwdhash, role);
    }

    public static UserNode create(String username, String name, String email, String password) {
        return create(username, name, email, password, "");
    }

    public UserNode newSession() {
        session = new BigInteger(130, RandUtils.secRand()).toString(32);
        return this;
    }

    public boolean checkPassword(String candidate) {
        return BCrypt.checkpw(candidate, pwdhash);
    }

    public boolean checkSession(String candidate) {
        return session.equals(candidate);
    }

    @Override
    public String toString() {
        return name;
    }

    @Override
    public String raw() {
        return "type: " + "user<br />" + "username: " + username + "<br />"
                + "name: " + name + "<br />" + "role: " + role + "<br />"
                + "lastSeen: " + lastSeen + "<br />";
    }

    public String getUsername() {
        return username;
    }

    public String getName() {
        return name;
    }

    public String getEmail() {
        return email;
    }

    public String getPwdhash() {
        return pwdhash;
    }

    public String getSession() {
        return session;
    }

    public String getRole() {
        return role;
    }

    public long getSessionTs() {
        return sessionTs;
    }

    public long getLastSeen() {
        return lastSeen;
    }
}